mag/mag-als31300.c

```diff
@@ -0,0 +1,413 @@
+
+#define LOG_TAG       "mag-als31300: "
+#define log(fmt, ...) printf(LOG_TAG fmt "\n", ##__VA_ARGS__)
+
+
+#define I2C_DEV_ADDR 0x60u
+#define CFG_REG1     0x02u
+#define DATA_REG1    0x28u
+#define DATA_REG2    0x29u
+
+#define CFG_REG1_I2C_ADDR_MASK    0x0001FC00u
+#define CFG_REG1_I2C_THRESH       0x00000200u
+#define CFG_REG1_ENABLE_Z         0x00000100u
+#define CFG_REG1_ENABLE_Y         0x00000080u
+#define CFG_REG1_ENABLE_X         0x00000040u
+#define CFG_REG1_CUSTOMER_EE_MASK 0x0000001Fu
+
+/* 12-bit signed data */
+#define DATA_SIGN_BIT           0x0800u
+#define DATA_GAUSS_MULTIPLIER_X 1 /* ALS31300EEJASR-JOY: 1 LSB/G */
+#define DATA_GAUSS_MULTIPLIER_Y 1 /* ALS31300EEJASR-JOY: 1 LSB/G */
+#define DATA_GAUSS_MULTIPLIER_Z 4 /* ALS31300EEJASR-JOY: 0.25 LSB/G */
+#define DATA_TEMP_MULTIPLIER    302
+#define DATA_TEMP_SUBTRACTOR    1702
+
+#define GAUSS_TO_MILITESLA_DIVIDER 10 /* 1 G = 0.1 mT */
+
+
+/* handling device file requests */
+#define DEV_ID_X    0
+#define DEV_ID_Y    1
+#define DEV_ID_Z    2
+#define DEV_ID_TEMP 3
+
+static const struct {
+	const char *dev;
+	int id;
+} DEVICES[] = {
+	{ "magX", DEV_ID_X },
+	{ "magY", DEV_ID_Y },
+	{ "magZ", DEV_ID_Z },
+	{ "magTemp", DEV_ID_TEMP },
+};
+
+
+struct mag_data_s {
+	/* values in microTesla (0.001 miliTesla) */
+	int32_t x;
+	int32_t y;
+	int32_t z;
+
+	/* value in miliCelsius */
+	int32_t temp;
+};
+
+
+static int read32(uint8_t addr, uint32_t *valOut)
+{
+	uint8_t val[4] = {};
+	if (i2c_regRead(I2C_DEV_ADDR, addr, val, sizeof(val)) < 0) {
+		log("failed to read register 0x%02x", addr);
+		return -1;
+	}
+
+	*valOut = (val[0] << 24) | (val[1] << 16) | (val[2] << 8) | val[3];
+	return 0;
+}
+
+
+static int getData(struct mag_data_s *data)
+{
+	int16_t rawX, rawY, rawZ, rawTemp;
+	uint32_t reg;
+
+	if (read32(DATA_REG1, &reg) < 0) {
+		return -1;
+	}
+
+	rawX = (reg >> 20) & 0x0FF0;
+	rawY = (reg >> 12) & 0x0FF0;
+	rawZ = (reg >> 4) & 0x0FF0;
+	rawTemp = (reg & 0x3F) << 6;
+
+	if (read32(DATA_REG2, &reg) < 0) {
+		return -1;
+	}
+
+	rawX |= ((reg >> 16) & 0x0F);
+	rawY |= ((reg >> 12) & 0x0F);
+	rawZ |= ((reg >> 8) & 0x0F);
+	rawTemp |= (reg & 0x3F);
+
+	rawX = (rawX ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+	rawY = (rawY ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+	rawZ = (rawZ ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+
+	/* EM field in microTesla */
+	data->x = 1000 * (int32_t)rawX * DATA_GAUSS_MULTIPLIER_X / GAUSS_TO_MILITESLA_DIVIDER;
+	data->y = 1000 * (int32_t)rawY * DATA_GAUSS_MULTIPLIER_Y / GAUSS_TO_MILITESLA_DIVIDER;
+	data->z = 1000 * (int32_t)rawZ * DATA_GAUSS_MULTIPLIER_Z / GAUSS_TO_MILITESLA_DIVIDER;
+
+	/* temperature in miliCelsius */
+	data->temp = 1000 * DATA_TEMP_MULTIPLIER * (rawTemp - DATA_TEMP_SUBTRACTOR) / 4096;
+
+	return 0;
+}
+
+
+/* verify CFG1 contents to ensure we have ALS31300 under I2C_DEV_ADDR */
+static int verifyCfgReg(uint32_t val)
+{
+	if ((val & CFG_REG1_CUSTOMER_EE_MASK) != 0) {
+		return -1; /* we don't store anything in customer EEPROM - should be 0 */
+	}
```

Comment:
Kinda risky, write would brick it permanently (until driver update, that is)

Reply:
intentional, right now support for writing EEPROM is not planned, If the feature would be implemented I expect the implementer to tackle the issue fully.

Side note: part of the configuration is also stored in eeprom (like enabled axes / 1v8 i2c compatibility) so any configuration change would need chnages to identification process. Unfortunately there is no const register we might enquiry:
![image](https://github.com/phoenix-rtos/phoenix-rtos-devices/assets/3939983/6826dd73-a3ac-40e9-9dfc-8dab935fa23c)

Reply:
Ack